[iOS] Support for layout container (Flow + AreaGrid)

wyAbhishek opened the pull request:
# Support for Flow Layout Container

- This PR adds support for the Layout container for elements of type AdaptiveCard, Container, Column and Table Cell.  
- This PR introduces a feature flag resolver interface that allows the host to implement and register it, enabling control of features through ECS settings.
- For flow layout container, we are writing equal width distribution logic for elements like Container, ColumnSet and TableCell.

A new **_layouts_** property is introduced on container elements. The value is an array of Layout objects. The layout objects look like this:

<img width="647" alt="Screenshot 2024-08-23 at 5 03 23 PM" src="https://github.com/user-attachments/assets/0920ea97-0c88-404f-811d-1c418ce609bf">

The layout to apply is determined the following way: 

- Look for the first layout in the list whose targetWidth property matches the current host width. If one is found, apply it 

- Otherwise, look for a layout that doesn’t have a targetWidth defined. If one is found, apply it 

- Otherwise, apply a default Layout.Stack layout and ignore other layouts in the list 

# Spec for Layout.Flow

<img width="653" alt="Screenshot 2024-08-23 at 4 57 38 PM" src="https://github.com/user-attachments/assets/f2039eeb-5034-49c9-aa9d-5263958c73fb">

# Spec for AreaGrid layout:

![Screenshot 2024-08-26 at 12 03 21 PM](https://github.com/user-attachments/assets/b616e3c8-b997-4cad-a1cb-fa435d720736)

# Sample Card:

## Flow Layout Container

![Simulator Screenshot - iPhone 15 Pro - 2024-08-26 at 16 45 54](https://github.com/user-attachments/assets/69b21ff3-742e-4995-ae38-8feb03cca346)

## Area Grid Container
![Simulator Screenshot - iPhone 15 Pro - 2024-08-28 at 13 08 41](https://github.com/user-attachments/assets/6017c02a-ffcf-4b1b-8fda-4eb1faaad815)

GabrielMedAlv commented:
Looks good!. Is there any test you consider could be added?. If some recursive layout card could be added (if not already there) to ensure working on that scenario too would be great.

wyAbhishek commented:
> Looks good!. Is there any test you consider could be added?. If some recursive layout card could be added (if not already there) to ensure working on that scenario too would be great.

Done

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRSVGIconHoldingView.mm:
```diff
@@ -56,20 +56,20 @@ - (void)setUpView
                                   constant:0];
-    trailing.priority = 499;
-    leading.priority = 499;
+//    trailing.priority = 499;
```
should this be removed?

hggz commented:
Thank you @wyAbhishek , this is a huge undertaking and changeset. Some feedback. Feel free to prioritize on what would be too much of a lift that doesn't make sense to have now to get this feature set out.

# Swift PM Support:

Your new header files added which are exposed to the swift package manager imports need to be symlinked to the iOS Include directory like done here:

https://github.com/microsoft/Teams-AdaptiveCards-Mobile/pull/155/files

in order to do this, go to the include directory in your terminal and run commands on the relevant header files:

IE: 

`ln -s ../ACRIFeatureFlagResolver.h ./ACRIFeatureFlagResolver.h`


# c_str() nullability
The conversion from c_str() to NSString using NSString stringWithCString:encoding: assumes the conversion will always succeed. 

Please add error handling to manage cases where the conversion might fail (e.g., if the string is nullptr or if there's a failure in conversion).

# No Default Case Handling in Switch Statements

The switch statements in methods like distribute: and shouldUseNewLayoutForView: do not have a default case. Add a default case to handle unexpected CardElementType values please.

# ACRView - Use of Magic Numbers for Width and Height

please turn magic numbers like -1 into a constant for us to track.

# ACRFlowLayout.h - lack of docs

add docs on this and other new additions if possible.

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ARCGridViewLayout.mm:
```diff
@@ -0,0 +1,369 @@
+                         row:area->GetRow()
+                      column:area->GetColumn()
+                    rowSpan:area->GetRowSpan()
+                 columnSpan:area->GetColumnSpan()];
```
let's align colons here.

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRRenderer.mm:
```diff
@@ -226,4 +291,66 @@ + (UIView *)render:(UIView<ACRIContentHoldingView> *)view
+        }
+    }
+
+    //    view.accessibilityElements = [(ACRContentStackView *)view getArrangedSubviews];
```
can we removed commented code?

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ARCGridViewLayout.mm:
```diff
@@ -0,0 +1,369 @@
+            [NSLayoutConstraint activateConstraints:@[
+                [rowView.topAnchor constraintEqualToAnchor:previousRow.bottomAnchor constant:_rowSpacing],
+                [rowView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
+                [rowView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor]
```
can we follow objective c code style - move open braces to new line, else statements on new line

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm:
```diff
@@ -156,7 +157,8 @@ - (UIView *)render:(UIView<ACRIContentHoldingView> *)viewGroup
+        NSString *areaName = [NSString stringWithCString:elem->GetAreaGridName()->c_str() encoding:NSUTF8StringEncoding];
+        [viewGroup addArrangedSubview:[viewGroup addPaddingFor:contentholdingview] withAreaName:areaName];
     }
 
```
here and everywhere let's follow the objective c style - open braces on new line.

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRContentHoldingUIView.mm:
```diff
@@ -156,14 +156,20 @@ - (void)layoutSubviews
+                if([container isKindOfClass:[ACRContentStackView class]])
+                {
+                    [(ACRContentStackView *)_viewGroup invalidateIntrinsicContentSize];
+                }
```
nit: space after if.

rahulpinjani1 commented on source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/ACRCustomFeatureFlagResolver.m:
```diff
@@ -0,0 +1,44 @@
+
+- (NSString *)stringForFlag:(NSString *)flag { 
+    return nil;
+}
```
lets annote nullable to the return type for methods returning nil

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRContainerRenderer.mm:
```diff
@@ -55,11 +109,19 @@ - (UIView *)render:(UIView<ACRIContentHoldingView> *)viewGroup
+    } else {
+        [ACRRenderer render:container
+                   rootView:rootView
+                     inputs:inputs
```
here and everywhere - else and else if on new line and opening braces on a new line.

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRLayoutHelper.mm:
```diff
@@ -0,0 +1,260 @@
+                       forElement:(std::shared_ptr<BaseCardElement> const &)elem
+                    andHostConfig:(ACOHostConfig *)config
+{
+    switch (elem->GetElementType())
```
lets check for if elementType is present.

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRLayoutHelper.mm:
```diff
@@ -0,0 +1,260 @@
+    }
+
+    BOOL isGridLayoutEnabled = [featureFlagResolver boolForFlag:@"isGridLayoutEnabled"] ?: NO;
+
```
this can be combined into one check.

wyAbhishek replied:
Trying to provide explicit pattern for getting values from FeatureFlagResolver. This way we always provide a default value, even if FeatureFlagResolver is nil or it doesn't return a value. 
Although, in case of BOOL it may look redundant, but I would prefer this redundancy over missing default case.  This patter will make sense for other types like NSString or NSArray. 
Please see if make sense.

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRRenderer.mm:
```diff
@@ -226,4 +291,66 @@ + (UIView *)render:(UIView<ACRIContentHoldingView> *)view
+            if (!elem->GetIsVisible())
+            {
+                continue;
+            }
```
These can be one combined into one statement.

wyAbhishek replied:
Kept this separate intentionally for better readability of code flow.  Old flow has different handling for visibility updates, this gives 1:1 match with code branching in old vs new render flow. 
Will you prefer it to be combined?

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRFlowLayout.mm:
```diff
@@ -0,0 +1,313 @@
+        return UIStackViewAlignmentFill;
+    }
+    
+    switch (_layout->GetHorizontalAlignment()) {
```
lets check if the alignment information is not nil/ is present.

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRFlowLayout.mm:
```diff
@@ -0,0 +1,313 @@
+    
+    NSLayoutConstraint *selfWidth = [self.widthAnchor constraintEqualToConstant:_availableRowSpace];
+    selfWidth.priority = 999;
+    NSLayoutConstraint *leading = [stackView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor constant:0];
```
since constant is set to 0 here we can use constraintEqualToAnchor and not specify the constant here.

rahulpinjani1 replied:
same below.

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRLayoutHelper.mm:
```diff
@@ -0,0 +1,260 @@
+        return;
+    }
+    std::vector<std::shared_ptr<BaseCardElement>> body = card->GetBody();
+    float childrenWidth = parentWidth/(body.size()) ;
```
lets check if body and size are present and not null

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRLayoutHelper.mm:
```diff
@@ -0,0 +1,260 @@
+            unsigned int spacing = getSpacing(elem->GetSpacing(), hostConfig);
+            long numberOfColumns = columnSetElem->GetColumns().size();
+            // removing padding of coloumSet and spacing between elemenets
+            float availableSpace = (parentWidth - (2 * padding) - ((numberOfColumns - 1)*spacing));
```
space after "*", ((numberOfColumns - 1) * spacing

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRLayoutHelper.mm:
```diff
@@ -0,0 +1,260 @@
+            std::vector<std::shared_ptr<TableColumnDefinition>> colums = table->GetColumns();
+            long numberOfColumns = colums.size();
+            int cellSpacing = [config getHostConfig]->GetTable().cellSpacing;
+            float availableSpace = (parentWidth - ((numberOfColumns - 1)*cellSpacing));
```
lets add space here- * cellSpacing

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRColumnView.mm:
```diff
@@ -23,7 +23,7 @@ - (void)config:(nullable NSDictionary<NSString *, id> *)attributes
     self.inputHandlers = [[NSMutableArray<ACRIBaseInputHandler> alloc] init];
 }
-- (void)addArrangedSubview:(UIView *)view
+- (void)addArrangedSubview:(UIView *)view withAreaName:(NSString*)areaName
```
space - NSString *

rahulpinjani1 commented on source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRColumnSetView.mm:
```diff
@@ -18,7 +18,7 @@ - (void)config:(nullable NSDictionary<NSString *, id> *)attributes
 }
 
-- (void)addArrangedSubview:(UIView *)view
+- (void)addArrangedSubview:(UIView *)view withAreaName:(NSString*)areaName
```
space - NSString *

wyAbhishek commented:
> Thank you @wyAbhishek , this is a huge undertaking and changeset. Some feedback. Feel free to prioritize on what would be too much of a lift that doesn't make sense to have now to get this feature set out.
> 
> # Swift PM Support:
> Your new header files added which are exposed to the swift package manager imports need to be symlinked to the iOS Include directory like done here:
> 
> https://github.com/microsoft/Teams-AdaptiveCards-Mobile/pull/155/files
> 
> in order to do this, go to the include directory in your terminal and run commands on the relevant header files:
> 
> IE:
> 
> `ln -s ../ACRIFeatureFlagResolver.h ./ACRIFeatureFlagResolver.h`
> 
> # c_str() nullability
> The conversion from c_str() to NSString using NSString stringWithCString:encoding: assumes the conversion will always succeed.
> 
> Please add error handling to manage cases where the conversion might fail (e.g., if the string is nullptr or if there's a failure in conversion).
> 
> # No Default Case Handling in Switch Statements
> The switch statements in methods like distribute: and shouldUseNewLayoutForView: do not have a default case. Add a default case to handle unexpected CardElementType values please.
> 
> # ACRView - Use of Magic Numbers for Width and Height
> please turn magic numbers like -1 into a constant for us to track.
> 
> # ACRFlowLayout.h - lack of docs
> add docs on this and other new additions if possible.

Thanks @hggz for review. 
I have addressed some items and left some for future PR. Following is the list:

1. Created symlink for ACRIFeatureFlagResolver
2. Added a wrapper method for optiona std::string to NSString converstion. This has all safety built in. 
3. Default handling is already there. Validated that it works. 
4. Not removing magic number, because we don't have a constant file. There are many magic number scattered through out the code, this will need separate PR
5. I will add documentation for new layouts. Some work is pending for phase1, I will take it up then.
